activestorage/lib/active_storage/service/gcs_service.rb

```diff
@@ -13,6 +13,23 @@ class Service::GCSService < Service
     class MetadataServerError < ActiveStorage::Error; end
     class MetadataServerNotFoundError < ActiveStorage::Error; end
 
+    class << self
+      def options_from_uri(uri:, query_hash: {}) # :nodoc:
+        sub_uri = uri.to_s.sub(/^gcs:\/\//, "")
+        credentials, rest = sub_uri.split("@", 2)
+        rest_parts = rest.split("/", 2)
+        project = rest_parts[0]
+        bucket = rest_parts[1].split("?").first
```

Comment:
This strikes me as more complicated than a single regexp with multiple named matches.